dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProverWithTracker.java

```diff
@@ -0,0 +1,185 @@
+    private final Set<String> declarations;
+
+    public ProverWithTracker(SolverContext ctx, String fileName, ProverOptions... options) {
+        this.fmgr = ctx.getFormulaManager();
+        this.prover = ctx.newProverEnvironment(options);
+        this.fileName = fileName;
+        this.declarations = new HashSet<>();
+        init();
+    }
+
+    // An empty filename means there is no need to dump the encoding
+    private boolean dump() {
+        return !fileName.isEmpty();
+    }
+    
+    private void init() {
+        if(dump()) {
+            try {
+                Files.deleteIfExists(Paths.get(fileName));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            StringBuilder description = new StringBuilder();
+            LocalDate currentDate = LocalDate.now();
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("YYYY-MM-DD");
+            description.append("Generated on: " + currentDate.format(formatter) + "\n");
+            description.append("Generator: Dartagnan\n");
+            description.append("Application: Bounded model checking for weak memory models\n");
+            description.append("Publications: \n" +
+                "- Hernán Ponce de León, Florian Furbach, Keijo Heljanko, " +
+                "Roland Meyer: Dartagnan: Bounded Model Checking for Weak Memory Models " +
+                "(Competition Contribution). TACAS (2) 2020: 378-382\n" +
+                "- Thomas Haas, Roland Meyer, Hernán Ponce de León: " +
+                "CAAT: consistency as a theory. Proc. ACM Program. Lang. 6(OOPSLA2): 114-144 (2022)"
+            );
+            write("(set-info :smt-lib-version 2.6)\n");
+            write("(set-logic ALL)\n");
+            write("(set-info :category \"industrial\")\n");
+            write("(set-info :source |\n" + description + "\n|)\n");
+            write("(set-info :license \"https://creativecommons.org/licenses/by/4.0/\")\n");
+        }
+    }
+
+    @Override
+    public void close() {
+        if(dump()) {
+            removeDuplicatedDeclarations(fileName);
+            write("(exit)\n");
+        }
+        prover.close();
+    }
+
+    public Void addConstraint(BooleanFormula f) throws InterruptedException {
+        if(dump()) {
+            write(fmgr.dumpFormula(f).toString());
+        }
+        return prover.addConstraint(f);
+    }
+
+    public boolean isUnsatWithAssumptions(Collection<BooleanFormula> fs) throws SolverException, InterruptedException {
+        long start = System.currentTimeMillis();
+        boolean result = prover.isUnsatWithAssumptions(fs);
+        long end = System.currentTimeMillis();
+        if(dump()) {
+            write("(push)\n");
+            for(BooleanFormula f : fs) {
+                write(fmgr.dumpFormula(f).toString());
+            }
+            write("(set-info :status " + (result ? "unsat" : "sat") + ")\n");
+            write("(check-sat)\n");
```

Comment:
The `status` this is not needed for debugging, but I am planning to eventually submit some of the encodings to SMT-COMP and there it is mandatory (including their ordering).

Because I need to put the status before the `check-sat`, I can only put up to the `push` before starting solver in dat3m side. If we end up with an scenario where really the smt solver does not finish, we just need to manually add a final `check-sat` and try other solver. However this scenario will not be that often. One of my main goals is to have files where dat3m eventually finished and being able to identify which iterations were slow and hopefully get some insight from smt solver developers.